### PR TITLE
Complete the build status with warnings and errors

### PR DESCRIPTION
**Pull request by flomillot**
Add two new build status to inform the user if a warning or an error happened during the building of a node

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java [hunk 1]
```diff
@@ -7,11 +7,31 @@
...
+            return BuildStatus.BUILT_WITH_ERROR;
+        } else if (status.equals(NetworkModificationResult.ApplicationStatus.WITH_WARNINGS)) {
+            return BuildStatus.BUILT_WITH_WARNING;
+        } else {
+            return BuildStatus.BUILT;
+        }
+    }
+
+    public boolean isBuilt() {
+        return ordinal() >= 2;
```
Comment (laurgarn) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java:35]: Did not verify, but could we use BUILT.ordinal() in stead of plain 2 ?

**Review by laurgarn — APPROVED** [review]
Code OK

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java (hunk 1 above)
Comment (flomillot) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java:35]: Good idea, I change it

File: src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java [hunk 2]
```diff
@@ -25,13 +25,15 @@ private ImpactUtils() {
...
-            .build());
+                .networkImpacts(List.of(createElementImpact(impactType, elementType, elementId, substationIds)))
+                .applicationStatus(NetworkModificationResult.ApplicationStatus.ALL_OK)
+                .build());
     }
 
     public static Optional<NetworkModificationResult> createModificationResultWithElementImpact(SimpleElementImpact impact) {
         return Optional.of(NetworkModificationResult.builder()
             .networkImpacts(List.of(impact))
+            .applicationStatus(NetworkModificationResult.ApplicationStatus.ALL_OK)
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java:36]: Not really needed, with Builder.Default annotation on applicationStatus field

File: src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java [hunk 3]
```diff
@@ -25,13 +25,15 @@ private ImpactUtils() {
 
     public static Optional<NetworkModificationResult> createModificationResultWithElementImpact(SimpleImpactType impactType, IdentifiableType elementType, String elementId, Set<String> substationIds) {
         return Optional.of(NetworkModificationResult.builder()
-            .networkImpacts(List.of(createElementImpact(impactType, elementType, elementId, substationIds)))
-            .build());
+                .networkImpacts(List.of(createElementImpact(impactType, elementType, elementId, substationIds)))
+                .applicationStatus(NetworkModificationResult.ApplicationStatus.ALL_OK)
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java:29]: Not really needed, with Builder.Default annotation on applicationStatus field

File: src/test/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatusTest.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        assertTrue(BUILT_WITH_WARNING.isBuilt());
+        assertTrue(BUILT.isBuilt());
+        assertFalse(BUILDING.isBuilt());
+        assertFalse(NOT_BUILT.isBuilt());
+    }
+
+    // If you change the values, then you will have to edit this test and be careful about the order of the enum as said in the BuildStatus class
+    @Test
+    void checkEnumValues() {
+        assertArrayEquals(values(), List.of(NOT_BUILT, BUILDING, BUILT, BUILT_WITH_WARNING, BUILT_WITH_ERROR).toArray());
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatusTest.java:32]: BuildStatus.values() would be less confusing ...

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java (hunk 2 above)
Comment (flomillot) [reply src/test/java/org/gridsuite/study/server/utils/ImpactUtils.java:36]: Indeed I forgot to remove it after adding Builder.Default

File: src/test/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatusTest.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package org.gridsuite.study.server.networkmodificationtree.dto;
```
[the remark this answers is not in the document]
Comment (flomillot) [reply src/test/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatusTest.java:1]: Done

File: src/test/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatusTest.java (hunk 4 above)
Comment (flomillot) [reply src/test/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatusTest.java:32]: Ok done

**Review by FranckLecuyer — APPROVED** [review]

File: src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java [hunk 6]
```diff
@@ -712,12 +713,33 @@ private void invalidateNodeProper(NodeEntity child, InvalidateNodeInfos invalida
     public void updateBuildStatus(UUID nodeUuid, BuildStatus buildStatus) {
         List<UUID> changedNodes = new ArrayList<>();
         UUID studyId = getStudyUuidForNodeId(nodeUuid);
+        NodeEntity nodeEntity = getNodeEntity(nodeUuid);
 
-        nodesRepository.findById(nodeUuid).ifPresent(n -> repositories.get(n.getType()).updateBuildStatus(nodeUuid, buildStatus, changedNodes));
+        BuildStatus newNodeStatus;
+        if (buildStatus.isBuilt()) {
+            NodeEntity previousBuiltNode = doGetLastParentNodeBuilt(nodeEntity);
+            BuildStatus previousBuiltNodeStatus = repositories.get(previousBuiltNode.getType()).getBuildStatus(previousBuiltNode.getIdNode());
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java:721]: Add a method with an explicit name to calculate the max

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java (hunk 6 above)
Comment (flomillot) [reply src/main/java/org/gridsuite/study/server/service/NetworkModificationTreeService.java:721]: Done

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java [hunk 7]
```diff
@@ -32,6 +35,10 @@ public static BuildStatus fromApplicationStatus(NetworkModificationResult.Applic
         }
     }
 
+    public static BuildStatus returnHigherSeverityStatus(BuildStatus buildStatus1, BuildStatus buildStatus2) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java]: Perhaps better to use an attribute severityLevel like `ApplicationStatus` and a method max
This solution does not depend of a java implementation of enum (ordinal) 

    NOT_BUILT (-1),
    BUILDING (-1),
    BUILT (0),
    BUILT_WITH_WARNING (1),
    BUILT_WITH_ERROR (2);

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java [hunk 8]
```diff
@@ -32,6 +35,10 @@ public static BuildStatus fromApplicationStatus(NetworkModificationResult.Applic
...
+        return Collections.max(List.of(buildStatus1, buildStatus2));
+    }
+
     public boolean isBuilt() {
         return ordinal() >= BUILT.ordinal();
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java:43]: return this.severityLevel >= 0
Do not use ordinal

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java (hunk 7 above)
Comment (flomillot) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java]: Yes I'm doing it

Comment (flomillot) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java]: Done

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java (hunk 8 above)
Comment (flomillot) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/BuildStatus.java]: Done

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**